src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1890 @@
+        tuples {Ri^TgmRj}, {Ri^TglRk} and {Rj^TgrRk}. Compute norms of the form
+        ||Ri^TgmRj*Rj^TglRk-Ri^TglRk||, ||J*Ri^TgmRj*J*Rj^TglRk-Ri^TglRk||,
+        ||Ri^TgmRj*J*Rj^TglRk*J-Ri^TglRk| and ||Ri^TgmRj*Rj^TglRk-J*Ri^TglRk*J||
+        where gm,gl,gr are the varipus gorup members of Dn and J=diag([1,1-1]).
+        The correct "J-configuration" is given for the smallest of these 4 norms.
+
+        :param Rijs: (n-choose-2)x3x3 array of relative rotations.
+        :return: List of n-choose-3 indices in {0,1,2,3} indicating
+            which J-configuration for each triplet of Rijs, i<j<k.
+        """
+        J_list = np.zeros(len(self.triplets), dtype="int")
+        R = Rijs
+        trip_idx = 0
+        pbar = tqdm(
+            desc="Finding best J-configuration over triplets", total=len(self.triplets)
+        )
+        for i, j, k in self.triplets:
+            ij = self.pairs_to_linear[i, j]
+            jk = self.pairs_to_linear[j, k]
+            ik = self.pairs_to_linear[i, k]
+            Rij, Rjk, Rik = R[ij], R[jk], R[ik]
+            Rjk_t = np.transpose(Rjk, (0, 2, 1))
+
+            Rij_J = J_conjugate(Rij)
+            Rjk_t_J = J_conjugate(Rjk_t)
+            Rik_J = J_conjugate(Rik)
+
+            final_votes = np.zeros(4)
+            final_votes[0] = self._compare_rots(Rij, Rjk_t, Rik)
+            final_votes[1] = self._compare_rots(Rij_J, Rjk_t, Rik)
+            final_votes[2] = self._compare_rots(Rij, Rjk_t_J, Rik)
+            final_votes[3] = self._compare_rots(Rij, Rjk_t, Rik_J)
+
+            J_list[trip_idx] = np.argmin(final_votes)
+            trip_idx += 1
+
+            pbar.update()
+        pbar.close()
+
+        return J_list
+
+    def _compare_rots(self, Rij, Rjk_t, Rik):
+        """
+        Compute norms for the 4 J-configurations and return indices
+        corresponding to best configuration for the provided triplet
+        of relative rotations.
+
+        :param Rij: Relative rotation between i'th and j'th candidate rotations
+            of shape (4, 3, 3).
+        :param Rjk_t: Transpose of relative rotation between j'th and k'th candidate
+            rotations of shape (4, 3, 3).
+        :param Rik: Relative rotation between i'th and k'th candidate rotations
+            of shape (4, 3, 3).
+        :return: Score for this J-configuration of the given rotation triplet.
+        """
+        # We compute the four sets of 4^3 norms |Rik @ Rjk.T - Rij|
+        # See equation (6.11) in publication.
+        prod_arr = Rik[:, None] @ Rjk_t[None]
+        diff_arr = prod_arr[:, :, None] - Rij
+        diff_arr = diff_arr.reshape((64, 9))
+        norm_arr = np.sum(diff_arr**2, axis=1)
+
+        # For perfect estimates, 16 of the 64 norms will equal zero.
+        # We sum over the smallest 16 values to get a vote for this J-configuration.
+        m = np.sort(norm_arr)
+        vote = np.sum(m[:16])
+
+        return vote
+
+    def _J_sync_power_method(self, J_list):
```

Comment:
This method is essentially the same as the ones in sync3n and C2. The only difference being, here I choose to always return the J_sync eigenvector with a positive first entry. That doesn't actually matter since it's arbitrary which set of rotations we decide to J-conjugate. I only did it to help match the matlab outputs from run to run.

This can probably lifted to the base class. Probably better to do while there is not multiple common line PRs out there.

Reply:
Yeah of course. Something for later then.

Reply:
Maybe we make an issue (if you don't already have one).

Reply:
Sure. Done: #1175 

Thanks.